File: 851.py
from typing import List
from collections import deque


class Solution:
    def loudAndRich(self, richer: List[List[int]], quiet: List[int]) -> List[int]:
        n = len(quiet)
        g = [[] for _ in range(n)]
        for a, b in richer:
            g[b].append(a)

        ans = [-1]*n

        def dfs(i):
            if ans[i] != -1:
                return
            ans[i] = i
            for j in g[i]:
                dfs(j)
                if quiet[ans[j]] < quiet[ans[i]]:
                    ans[i] = ans[j]

        for i in range(n):
            dfs(i)
        return ans

    def TopologicalSorting(self, richer: List[List[int]], quiet: List[int]) -> List[int]:
        n = len(quiet)
        inDegree = [0]*n
        g = [[] for _ in range(n)]
        for a, b in richer:
            g[a].append(b)
            inDegree[b] += 1

        ans = list(range(n))
        q = deque((i for i in range(n) if inDegree[i] == 0))
        while q:
            x = q.popleft()
            for y in g[x]:
                if quiet[ans[x]] < quiet[ans[y]]:
                    ans[y] = ans[x]
                inDegree[y] -= 1
                if inDegree[y] == 0:
                    q.append(y)
        return ans

File: 851.go
package leetcode

func loudAndRich(richer [][]int, quiet []int) []int {
	n := len(quiet)
	edges := make([][]int, n)
	rich := make([]bool, n)
	for _, e := range richer {
		edges[e[1]] = append(edges[e[1]], e[0])
		rich[e[0]] = true
	}

	ans := make([]int, n)
	visit := map[int][2]int{}

	var helper func(x int) (int, int)
	helper = func(x int) (min_q, min_p int) {
		if val, ok := visit[x]; ok {
			return val[0], val[1]
		}
		// visit[x] = [2]int{-1, -1}
		min_q = quiet[x]
		min_p = x
		for _, y := range edges[x] {
			q, p := helper(y)
			if q < min_q {
				min_q = q
				min_p = p
			}
		}
		ans[x] = min_p
		// visit[x][0] = min_q
		// visit[x][1] = min_p
		visit[x] = [2]int{min_q, min_p}
		return
	}

	for p, isRich := range rich {
		if !isRich {
			helper(p)
		}
	}
	return ans
}
